### PR TITLE
GEOMESA-1997 Enrichment services in converter framework

### PR DESCRIPTION
**Pull request by anthonyccri**
* Added enrichment caches to converter framework to load
  static data via a lookup key
* Implemented a Redis version of the lookup cache

Signed-off-by: Anthony Fox <anthony.fox@ccri.com>

### CODE REVIEW

File: docs/user/convert/function_usage.rst [hunk 1]
```diff
@@ -567,3 +567,16 @@ delimiters for a map:
 ::
 
     { name = "numbers", transform = "parseMap('int -> string', $2, ',', '->')" }
+
+
+Enrichment Functions
+~~~~~~~~~~~~~~~~~~~~
+
+cacheLookup
+^^^^^^^^
```
Comment (elahrvivaz) [inline docs/user/convert/function_usage.rst]: carrots need to match length of title to be rendered correctly

File: geomesa-convert/geomesa-convert-redis-cache/pom.xml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+    <dependencies>
+        <dependency>
+            <groupId>org.locationtech.geomesa</groupId>
+            <artifactId>geomesa-convert-common_${scala.binary.version}</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>redis.clients</groupId>
+            <artifactId>jedis</artifactId>
+            <version>2.9.0</version>
```
Comment (elahrvivaz) [inline geomesa-convert/geomesa-convert-redis-cache/pom.xml]: we'll need to create cqs for this - can you run `./build/calculate-cqs.sh` and push up the modified `./build/cqs.tsv`?

File: geomesa-convert/geomesa-convert-redis-cache/src/main/scala/org/locationtech/geomesa/convert/redis/RedisEnrichmentCache.scala [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    }
+    new RedisEnrichmentCache(connBuilder, url, timeout)
+  }
+}
+
+
+object RedisEnrichmentCache {
+
+  case class LookupRedis() extends TransformerFn {
+    override def names: Seq[String] = Seq("lookupRedis")
```
Comment (elahrvivaz) [inline geomesa-convert/geomesa-convert-redis-cache/src/main/scala/org/locationtech/geomesa/convert/redis/RedisEnrichmentCache.scala]: doesn't look like this is registered anywhere - is it redundant with 'cacheLookup' and can be removed?

File: geomesa-convert/geomesa-convert-common/src/main/scala/org/locationtech/geomesa/convert/SimpleFeatureConverterFactory.scala [hunk 4]
```diff
@@ -257,6 +272,7 @@ trait ToSimpleFeatureConverter[I] extends SimpleFeatureConverter[I] with LazyLog
   def inputFields: Seq[Field]
   def idBuilder: Expr
   def userDataBuilder: Map[String, Expr]
+  def caches: Map[String, EnrichmentCache]
```
Comment (elahrvivaz) [inline geomesa-convert/geomesa-convert-common/src/main/scala/org/locationtech/geomesa/convert/SimpleFeatureConverterFactory.scala]: maybe we should provide a default implementation here of `= Map.empty` to not break other `ToSimpleFeatureConverter` impls, or do we not care/want to force them to implement the caching?

**Comment by elahrvivaz** [comment]
looks like your CLA is under `anthonyfox@ccri.com` but you're using `anthony.fox@ccri.com` here

File: geomesa-convert/geomesa-convert-redis-cache/src/main/scala/org/locationtech/geomesa/convert/redis/RedisEnrichmentCache.scala (hunk 3 above)
Comment (anthonyccri) [reply geomesa-convert/geomesa-convert-redis-cache/src/main/scala/org/locationtech/geomesa/convert/redis/RedisEnrichmentCache.scala]: Yes, I will remove it.

File: geomesa-convert/geomesa-convert-common/src/main/scala/org/locationtech/geomesa/convert/SimpleFeatureConverterFactory.scala (hunk 4 above)
Comment (anthonyccri) [reply geomesa-convert/geomesa-convert-common/src/main/scala/org/locationtech/geomesa/convert/SimpleFeatureConverterFactory.scala]: Hmm, we could provide a default impl but then if a custom impl went to use the caches, they would have to hunt this down.  I could see it going both ways.

File: build/cqs.tsv [hunk 5]
```diff
@@ -225,6 +225,7 @@ org.slf4j:slf4j-api	1.7.5	compile
 org.slf4j:slf4j-log4j12	1.7.5	compile
 org.tukaani:xz	1.0	compile
 picocontainer:picocontainer	1.2	compile
+redis.clients:jedis	2.9.0	compile
```
Comment (elahrvivaz) [inline build/cqs.tsv]: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=14037

**Comment by jnh5y** [comment]
Uhmm... we haven't hit ticket 9997:)...

File: pom.xml [hunk 6]
```diff
@@ -77,7 +77,7 @@
         <thrift.version>0.9.1</thrift.version>
         <hadoop.version>2.5.2</hadoop.version>
         <spark.version>2.0.0</spark.version>
-        <hbase.version>1.2.3</hbase.version>
+        <hbase.version>1.3.1</hbase.version>
```
Comment (elahrvivaz) [inline pom.xml]: this will probably require cqs... maybe we should have this as a separate pr, and/or hold off on bumping hbase versions until GM 1.4?

Comment (jnh5y) [reply pom.xml]: I'd suggest CQs 'now', and hold the HBase updates until 1.4.0.  

Tentatively, we could start a '1.4.x' CQ-loose-kicking PR.

**Pull request merged**